File: docs/reference/OpenViking/openviking_cli/cli/commands/relations.py
# Copyright (c) 2026 Beijing Volcano Engine Technology Co., Ltd.
# SPDX-License-Identifier: Apache-2.0
"""Relation management commands."""

from typing import List

import typer

from openviking_cli.cli.context import get_cli_context
from openviking_cli.cli.errors import execute_client_command, run
from openviking_cli.cli.output import output_success


def register(app: typer.Typer) -> None:
    """Register relation commands."""

    @app.command("relations")
    def relations_command(
        ctx: typer.Context,
        uri: str = typer.Argument(..., help="Viking URI"),
    ) -> None:
        """List relations of a resource."""
        run(ctx, lambda client: client.relations(uri))

    @app.command("link")
    def link_command(
        ctx: typer.Context,
        from_uri: str = typer.Argument(..., help="Source URI"),
        to_uris: List[str] = typer.Argument(..., help="One or more target URIs"),
        reason: str = typer.Option("", "--reason", help="Reason for linking"),
    ) -> None:
        """Create relation links from one URI to one or more targets."""
        cli_ctx = get_cli_context(ctx)
        result = execute_client_command(
            cli_ctx,
            lambda client: client.link(from_uri, to_uris, reason),
        )
        output_success(
            cli_ctx,
            result if result is not None else {"from": from_uri, "to": to_uris, "reason": reason},
        )

    @app.command("unlink")
    def unlink_command(
        ctx: typer.Context,
        from_uri: str = typer.Argument(..., help="Source URI"),
        to_uri: str = typer.Argument(..., help="Target URI to unlink"),
    ) -> None:
        """Remove a relation link."""
        cli_ctx = get_cli_context(ctx)
        result = execute_client_command(
            cli_ctx,
            lambda client: client.unlink(from_uri, to_uri),
        )
        output_success(
            cli_ctx,
            result if result is not None else {"from": from_uri, "to": to_uri},
        )

File: docs/reference/OpenViking/crates/ov_cli/src/commands/relations.rs
use crate::client::HttpClient;
use crate::error::Result;
use crate::output::{output_success, OutputFormat};

pub async fn list_relations(
    client: &HttpClient,
    uri: &str,
    format: OutputFormat,
    compact: bool,
) -> Result<()> {
    let result = client.relations(uri).await?;
    output_success(&result, format, compact);
    Ok(())
}

pub async fn link(
    client: &HttpClient,
    from_uri: &str,
    to_uris: &Vec<String>,
    reason: &str,
    format: OutputFormat,
    compact: bool,
) -> Result<()> {
    let result = client.link(from_uri, to_uris, reason).await?;
    // If the server returns null/empty, show a confirmation summary
    if result.is_null() {
        let summary = serde_json::json!({
            "from": from_uri,
            "to": to_uris,
            "reason": reason,
        });
        output_success(&summary, format, compact);
    } else {
        output_success(&result, format, compact);
    }
    Ok(())
}

pub async fn unlink(
    client: &HttpClient,
    from_uri: &str,
    to_uri: &str,
    format: OutputFormat,
    compact: bool,
) -> Result<()> {
    let result = client.unlink(from_uri, to_uri).await?;
    if result.is_null() {
        let summary = serde_json::json!({
            "from": from_uri,
            "to": to_uri,
        });
        output_success(&summary, format, compact);
    } else {
        output_success(&result, format, compact);
    }
    Ok(())
}
